proposed structure for time-series target-data (#22)
reichlab/decisions: decisions/2025-02-27-rfc-time-series-target-data.md

# Time-series target data proposal

## General idea/Aims/Requirements

* Prescribe a file-location and \-format for time-series target data that can be validated and used to support dashboard visualization and potentially could be used for other purposes, such as modeling.

> * The format for time-series target data will apply to targets that are defined as “step-ahead” targets in the target-metadata and have a numeric data type. \[Note: other data, including time-series data, could be stored in target-data but all files whose name or path starts with “time-series\*” (or possibly “oracle-output\*”, depending on how we end up setting that up) will be subject to validation.\]

Reviewer: Just to note that it might be good to clearly re-iterate that:
- a single file should be called `timeseries.{ext}`. 
- If the timeseries data is to be split across files it should be stored within a `target-data/time-series/` directory. Within that directory, filenames do not matter but files do need to all share the same file format (either all csv or all parquet). 
- Hive partitioning (i.e. directory names which take the form `column_name = value` is only required if the hub decides to reduce the size of it's data by encoding some of it into directory names. 
- But they could choose to just the data within the `time-series` directory into multiple files and even subdirectories that each contains all time-series columns and that will work just fine too. In this case, useing hive-style directory naming is not required.

Author: I was thinking that the single file would be
`target-data/time-series.{ext}`
and the partition would exist at
`target-data/time-series/`
but you are suggesting that there should be no hyphen in the file name? Why not keep them consistent?

Reviewer: My bad! That was a fast typing typo!

> * The format for time-series target data will apply to targets that are defined as “step-ahead” targets in the target-metadata and have a numeric data type. \[Note: other data, including time-series data, could be stored in target-data but all files whose name or path starts with “time-series\*” (or possibly “oracle-output\*”, depending on how we end up setting that up) will be subject to validation.\]

Author: I have addressed this in an update.

> * We will support single files and hive-style partitions where the partition is based on the as\_of date. Any time data are versioned using as “as\_of” partition or column, the full version must be made available, no support for “diffs” from previous versions will be supported.

Reviewer: Do you mean `version` instead of `as_of` here?

Author: We have decided to migrate back to using "as_of" as the column name, as this is more consistent with the Delphi epidata API conventions. I will add a justification for this in the file.

Author: https://cmu-delphi.github.io/epiprocess/articles/archive.html

* All time-series target data files must either be .csv or .parquet. No mixing of file types will be allowed.
* Validations are enforced based on properties of the target that can be read from the tasks.json file.
* A target can be represented in a time-series file as long as the definition of that target-id is consistent across all rounds/tasks. For example, as long as the names of the task-id variables are the same across all rounds, then it is easy to understand which columns should be present in a time-series file. If those column names change, then it becomes more complicated to support.

## Summary of the proposal

> - One time-series.csv (or .parquet) file or partition contains all of the validatable time-series target data for numeric variables in the hub.

Reviewer 2: ```suggestion
- One time-series.csv (or .parquet) file or partitioned data set contains all of the validatable time-series target data for numeric variables in the hub.
```

- This file has the following columns:
  - “as\_of”: optional, an ISO date that says when the data were available
  - \[a subset of task-id variables\]
    - these variables define the unit of observation for each observed value

>     - one task-id variable must be an ISO date

Reviewer 2: ```suggestion
    - one task-id variable must be an ISO date, corresponding to the date of the observed event
```

    - initially, they just need to be task-id variables, and we will later implement a metadata field/validation about which task-id variables these are.
    - they must be readable as the data type for the task-id variable in the schema
  - “observation”: a numeric observed value for the unit

> ## Key justifications
> 
> - We are preparing for a time when the dev team is smaller and therefore are trying to simplify the hub admin user experience. It has been noted that hub admin experience with hive partitions is limited and documentation for these partitions, even for the initiated, are complex.
> - By focusing our support on several key use-cases (e.g. only numeric target observations), we make it easier to support single-file target data. One possible simplification is that mandating a single file gets us out of schema mismatch issues later on (e.g. reading partition when observation column is interpolated as different data types).
> -

Reviewer 3: ```suggestion
```

I don't think we need this section anymore, since it was primarily about justifying the idea to drop partitions altogether.

Reviewer 2: agree

  - Example for a hub that tracks covid cases, hospitalizations, and deaths, recorded via the “target” task id variable:
    EXAMPLE 1: multiple targets
    target\_metadata: \[
        {
            “Target\_id”: “cases”,
             …,
            “Target\_keys”: {“target”: “cases”},
            …
        },
        {
            “Target\_id”: “hospitalizations”,
             …,
            “Target\_keys”: {“target”: “hospitalizations”},
            …
        },
        {
            “target\_id” : “deaths”,
             …,
            “Target\_keys”: {“target”: “deaths”},
            …
        }
    \]

Then the time-series.csv would look like this

| version | target | date | location | observation |
| :---- | :---- | :---- | :---- | :---- |
|  | cases |  |  |  |
|  | cases |  |  |  |
|  | … |  |  |  |
|  | deaths |  |  |  |
|  | deaths |  |  |  |
|  | … |  |  |  |

EXAMPLE 2: one target, with target\_keys: null
target\_metadata: \[
    {
        “target\_id”: “cases”,
         …,
        “target\_keys”: null
        …
    }
\]

## Validations

A file hub/target-data/time-series.csv (or .parquet, referred to below as .csv for simplicity) is valid if

> 1. all model-task blocks in tasks.json \> rounds with the named target-id in target-metadata have

Reviewer 4: Can someone explain what the ">" means in this line to me? i'm not super familiar with the schemas, so i might be missing something

Reviewer 2: It's being used here to indicate "the rounds property of the tasks.json file"

Author: I rephrased to make this clearer.

   1. is\_step\_ahead: true
   2. target\_type metadata field is one of “continuous” “discrete” “binary” or “compositional”
   3. the same task-id variable names
   4. the same task-id variables specified as what define the observable unit (Note: we can only perform this validation if we have a system for indicating which task-id variables define the observable unit. We will not implement this validation to start, but may add it later when schemas are updated.).
2. The column names of time-series.csv are
   1. “observation”
   2. The task-id column names that define the observable unit

>       1. in our initial implementation, these are defined by the hub admin and not validatable. At a future time, we could implement a metadata field to capture these so it is validatable.

Reviewer 2: ```suggestion
      1. in our initial implementation, these are defined by the hub admin and not validatable other than that they are task id variable names listed in the hub's `tasks.json` file. At a future time, we could implement a metadata field to capture these so it is validatable.
```

### Supporting multiple single files, one per target

It was agreed that this is the least accessible option. It would always require multiple bespoke reads of individual files, rather than taking advantage of the hive style partition structure or of just storing all the data in one place.

### Adding a new schema property to define the observed unit variables

The idea here would be to add a new schema property that defines which task-id variables correspond to the variables that define the observational unit. This mapping could also allow for renaming certain variables, e.g. task-id variable of “target\_end\_date” could just be “date” which would be more natural in the context of a time-series target data file.

This is something that we would like to do at some point, but doesn’t feel essential to making this work in the short term.

### Allowing for additional partition structure

It has been suggested that allowing for partitioning time-series target data by target might also be reasonable. One possible justification for this is operational: some targets come from different data sources and are updated at different times.

> This is also something that we could consider explicitly supporting at a future date. It is possible that this would “just work” using the initial implementation and open-dataset.

Reviewer: I can confirm that this already just works for hive-partitioned data as well as data split across multiple files within the `time-series` directory and tests for these are included in the draft `connect_target_timeseries()` PR in prep
